src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.StepListener;
+import org.opensearch.action.bulk.BulkItemResponse;
+import org.opensearch.action.bulk.BulkRequestBuilder;
+import org.opensearch.action.bulk.BulkResponse;
+import org.opensearch.action.support.master.AcknowledgedResponse;
+
+/**
+ * Uploader will upload GeoJSON objects from UploadGeoJSONRequestContent as
+ * Documents to given index in three stage.
+ * At first stage (preUpload), resources like index, mapping, pipeline with
+ * GeoJSON Feature processors, will be created.
+ * At second stage (upload), Feature will be extracted from GeoJSON and indexed using
+ * BulkAction. This supports both Feature and FeatureCollection.
+ * At third stage (postUpload), previously created pipeline will be deleted
+ * At final stage response or failure will be added to the listener.
+ */
+public class Uploader {
+
+    public static final String ERROR_MESSAGE_DELIMITER = "\n";
+
+    private static final Logger LOGGER = LogManager.getLogger(Uploader.class);
+
+    private final IndexManager indexManager;
+    private final PipelineManager pipelineManager;
+    private final ContentBuilder contentBuilder;
+
+    public Uploader(IndexManager indexManager, PipelineManager pipelineManager, ContentBuilder contentBuilder) {
+
+        this.indexManager = Objects.requireNonNull(indexManager, "IndexManager instance cannot be null");
+        this.pipelineManager = Objects.requireNonNull(pipelineManager, "PipelineManager instance cannot be null");
+        this.contentBuilder = Objects.requireNonNull(contentBuilder, "ContentBuilder instance cannot be null");
+    }
+
+    /**
+     * upload abstracts following operations from request
+     * 1. Create index if it doesn't exist.
+     * 2. Create pipeline with {@link org.opensearch.geospatial.processor.FeatureProcessor}
+     * 3. Prepare Content from {@link UploadGeoJSONRequestContent#getData()}
+     * 4. Upload content
+     * 5. Delete pipeline
+     * @param content {@link UploadGeoJSONRequestContent} derived from {@link UploadGeoJSONRequest}
+     * @param isIndexAlreadyExists confirms whether the uploader should create the new index or not
+     */
+    public void upload(
+        UploadGeoJSONRequestContent content,
+        boolean isIndexAlreadyExists,
+        ActionListener<AcknowledgedResponse> flowListener
+    ) {
+        // validate input
+        Objects.requireNonNull(flowListener, "listener cannot be null");
+        Objects.requireNonNull(content, "content cannot be null");
+
+        // initialize step listeners to chain steps
+        StepListener<Void> createIndexStep = new StepListener<>();
+        StepListener<String> createPipelineStep = new StepListener<>();
+        StepListener<Void> indexFeatureStep = new StepListener<>();
+        StepListener<Exception> deletePipelineStep = new StepListener<>();
+
+        if (isIndexAlreadyExists) {
```

Comment:
> which in turn doesn't trigger any more actions.

Does this mean that the call would fail? On resource already exists could we just continue to next step?

Reply:
createIndexStep will be set to Failure. This is the beginning of pipeline, if it fails, nothing else will be execute.